diffrax/brownian/tree.py

```diff
@@ -25,21 +25,75 @@
 # }
 #
 
+# We define
+# H_{s,t} = 1/(t-s) ( \int_s^t ( W_u - (u-s)/(t-s) W_{s,t} ) du ).
+# bhh_t = t * H_{0,t}
+# For more details see Definition 4.2.1 and Theorem 6.1.4 of
+#
+# Foster, J. M. (2020). Numerical approximations for stochastic
+# differential equations [PhD thesis]. University of Oxford.
+
 
 class _State(eqx.Module):
-    s: Scalar
-    t: Scalar
-    u: Scalar
-    w_s: Scalar
-    w_t: Scalar
-    w_u: Scalar
+    level: int
+    stu: tuple[Scalar, Scalar, Scalar]  # s, t, u
+    w_stu: tuple[Scalar, Scalar, Scalar]  # W at times s, t, u
+    w_st_tu: tuple[Scalar, Scalar]  # W_{s,t} and W_{t,u}
     key: "jax.random.PRNGKey"
+    bhh_stu: Optional[Tuple[Scalar, Scalar, Scalar]]  # \bar{H} at times s, t, u
+    bhh_st_tu: Optional[Tuple[Scalar, Scalar]]  # \bar{H}_{s,t} and \bar{H}_{t,u}
+    bkk_stu: Optional[Tuple[Scalar, Scalar, Scalar]]  # \bar{K} at times s, t, u
+    bkk_st_tu: Optional[Tuple[Scalar, Scalar]]  # \bar{K}_{s,t} and \bar{K}_{t,u}
+
+
+def _levy_diff(x0: LevyVal, x1: LevyVal) -> LevyVal:
+    r"""Computes $(W_{s,u}, H_{s,u})$ from $(W_s, \bar{H}_{s,u})$ and
+    $(W_u, \bar{H}_u)$, where $\bar{H}_u = u * H_u$.
+
+    **Arguments:**
+
+    - `x0`: `LevyVal` at time `s`
+
+    - `x1`: `LevyVal` at time `u`
+
+    **Returns:**
+
+    `LevyVal(W_su, H_su)`
+    """
+
+    su = (x1.dt - x0.dt).astype(x0.W.dtype)
+    w_su = x1.W - x0.W
+    if x0.H is None or x1.H is None:  # BM only case
+        return LevyVal(dt=su, W=w_su, H=None, bar_H=None, K=None, bar_K=None)
+
+    # levy_area == "space-time"
+    _su = jnp.where(jnp.abs(su) < jnp.finfo(su).eps, jnp.inf, su)
+    inverse_su = 1 / _su
+    u_bb_s = x1.dt * x0.W - x0.dt * x1.W
+    bhh_su = x1.bar_H - x0.bar_H - 0.5 * u_bb_s  # bhh_su = H_{s,u} * (u-s)
+    hh_su = inverse_su * bhh_su
+
+    return LevyVal(dt=su, W=w_su, H=hh_su, bar_H=None, K=None, bar_K=None)
+
+
+def split_interval(_cond, x_stu, x_st_tu):
```

Comment:
If this is a helper private to this file then begin it with an underscore: `_split_interval`.